chia/pools/pool_wallet.py

```diff
@@ -0,0 +1,857 @@
+    is_pool_member_inner_puzzle,
+    is_pool_waitingroom_inner_puzzle,
+    uncurry_pool_waitingroom_inner_puzzle,
+)
+
+from chia.util.ints import uint8, uint32, uint64
+from chia.wallet.derive_keys import (
+    find_owner_sk,
+    master_sk_to_pooling_authentication_sk,
+    master_sk_to_singleton_owner_sk,
+)
+from chia.wallet.transaction_record import TransactionRecord
+from chia.wallet.util.wallet_types import WalletType
+from chia.wallet.wallet import Wallet
+
+from chia.wallet.wallet_info import WalletInfo
+from chia.wallet.util.transaction_type import TransactionType
+
+
+class PoolWallet:
+    MINIMUM_INITIAL_BALANCE = 1
+    MINIMUM_RELATIVE_LOCK_HEIGHT = 10
+
+    wallet_state_manager: Any
+    log: logging.Logger
+    wallet_info: WalletInfo
+    target_state: Optional[PoolState]
+    standard_wallet: Wallet
+    wallet_id: int
+    """
+    From the user's perspective, this is not a wallet at all, but a way to control
+    whether their pooling-enabled plots are being self-farmed, or farmed by a pool,
+    and by which pool. Self-pooling and joint pooling rewards are swept into the
+    users' regular wallet.
+
+    If this wallet is in SELF_POOLING state, the coin ID associated with the current
+    pool wallet contains the rewards gained while self-farming, so care must be taken
+    to disallow joining a new pool while we still have money on the pooling singleton UTXO.
+
+    Pools can be joined anonymously, without an account or prior signup.
+
+    The ability to change the farm-to target prevents abuse from pools
+    by giving the user the ability to quickly change pools, or self-farm.
+
+    The pool is also protected, by not allowing members to cheat by quickly leaving a pool,
+    and claiming a block that was pledged to the pool.
+
+    The pooling protocol and smart coin prevents a user from quickly leaving a pool
+    by enforcing a wait time when leaving the pool. A minimum number of blocks must pass
+    after the user declares that they are leaving the pool, and before they can start to
+    self-claim rewards again.
+
+    Control of switching states is granted to the owner public key.
+
+    We reveal the inner_puzzle to the pool during setup of the pooling protocol.
+    The pool can prove to itself that the inner puzzle pays to the pooling address,
+    and it can follow state changes in the pooling puzzle by tracing destruction and
+    creation of coins associate with this pooling singleton (the singleton controlling
+    this pool group).
+
+    The user trusts the pool to send mining rewards to the <XXX address XXX>
+    TODO: We should mark which address is receiving funds for our current state.
+
+    If the pool misbehaves, it is the user's responsibility to leave the pool
+
+    It is the Pool's responsibility to claim the rewards sent to the pool_puzzlehash.
+
+    The timeout for leaving the pool is expressed in number of blocks from the time
+    the user expresses their intent to leave.
```

Comment:
I note that this might be a repetition of what's been said in the 6th paragraph.